### PR TITLE
support sync workspace to remote ssh host

### PR DESCRIPTION
**Pull request by yungezz**
## Summary

support sync workspace to remote ssh host
- add new command to sync workspace to ssh host
- add a warning message during run playbook via ssh, to ask for user whether to sync workspace.

## Issue Type

- New Feature

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -30,35 +30,41 @@
...
         "title": "Run Ansible Playbook in Local Ansible",
-        "category": "ansible"
+        "category": "Ansible"
       },
       {
         "command": "vscode-ansible.ssh",
         "title": "Run Ansible Playbook Remotely via ssh",
-        "category": "ansible"
+        "category": "Ansible"
       },
```
Comment (zikalino) [inline package.json]: probably these changes should be in a separate PR, but that's ok, as they are already there

File: src/fileSyncer.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import * as path from 'path';
+import { SSHServer } from './interfaces';
+import * as sshHelper from './sshRunner';
+
+export enum TargetType {
+    cloudshell = 'cloudshell',
+    remotehost = 'remotehost'
+}
+
+export class FileSyncer {
```
Comment (zikalino) [inline src/fileSyncer.ts]: I think FolderSyncer or WorkspaceSyncer would be better

File: src/fileSyncer.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            rootWorkspace = vscode.workspace.workspaceFolders[0].uri.fsPath;
+        }
+        vscode.window.showInputBox({ value: rootWorkspace, prompt: 'Please input folder path', placeHolder: 'folder', password: false })
+            .then((folder) => {
+                if (!folder) {
+                    return
+                }
+
+                if (!fs.existsSync(folder)) {
+                    this._outputChannel.appendLine('Folder not exists: ' + folder);
```
Comment (zikalino) [inline src/fileSyncer.ts]: Folder does not exist

File: src/cloudConsole.ts [hunk 4]
```diff
@@ -150,7 +150,7 @@ export async function openCloudConsole(api: AzureAccount, os: OS, files, outputC
 			} else {
 				for (let file of files) {
 					const data = fsExtra.readFileSync(file, { encoding: 'utf8' }).toString();
-					outputChannel.append('\nUpload playbook to CloudShell: ' + file + ' as $HOME/' + path.basename(file));
+					outputChannel.appendLine('\nUpload playbook to CloudShell: ' + file + ' as $HOME/' + path.basename(file));
```
Comment (zikalino) [inline src/cloudConsole.ts]: shouldn't we now have "upload *file* to CloudShell" as it may not necessarily be playbook?

File: src/fileSyncer.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    }
+
+    public syncFolderToRemoteSSHHost(): void {
+        let rootWorkspace = '';
+        if (vscode.workspace.workspaceFolders) {
+            rootWorkspace = vscode.workspace.workspaceFolders[0].uri.fsPath;
+        }
+        vscode.window.showInputBox({ value: rootWorkspace, prompt: 'Please input folder path', placeHolder: 'folder', password: false })
+            .then((folder) => {
```
Comment (zikalino) [inline src/fileSyncer.ts]: i think in VSC there's a specialized folder selector, so maybe you could try that?

File: src/fileSyncer.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+                        if (!server) {
+                            return;
+                        }
+
+                        this._outputChannel.append('Copying folder ' + folder + ' to ' + server.host);
+                        this._outputChannel.show();
+
+                        const progress = utilities.delayedInterval(() => { this._outputChannel.append('.') }, 500);
+                        utilities.copyFileRemote(folder, this.getTargetFolder(folder), server, (err) => {
```
Comment (zikalino) [inline src/fileSyncer.ts]: i think this function should be renamed, because it suggests it's copying single file

File: src/cloudConsole.ts (hunk 4 above)
Comment (yungezz) [reply src/cloudConsole.ts]: fixed. those part of code not necessary any more. we're not using azure storage uploading

File: src/fileSyncer.ts (hunk 2 above)
Comment (yungezz) [reply src/fileSyncer.ts]: renamed

**Comment by yungezz** [comment]
@zikalino resolved comments. and refined to use promise and async.
